crates/sui-indexer-alt/src/ingestion/remote_client.rs

```diff
@@ -0,0 +1,340 @@
+            code @ StatusCode::NOT_FOUND => {
+                debug!(checkpoint, %code, "Checkpoint not found");
+                Err(BE::permanent(Error::NotFound(checkpoint)))
+            }
+
+            // Timeouts are a client error but they are usually transient.
+            code @ StatusCode::REQUEST_TIMEOUT => Err(self.metrics.inc_retry(
+                checkpoint,
+                "timeout",
+                Error::HttpError(checkpoint, code),
+            )),
+
+            // Rate limiting is also a client error, but the backoff will eventually widen the
+            // interval appropriately.
+            code @ StatusCode::TOO_MANY_REQUESTS => Err(self.metrics.inc_retry(
+                checkpoint,
+                "too_many_requests",
+                Error::HttpError(checkpoint, code),
+            )),
+
+            // Assume that if the server is facing difficulties, it will recover eventually.
+            code if code.is_server_error() => Err(self.metrics.inc_retry(
+                checkpoint,
+                "server_error",
+                Error::HttpError(checkpoint, code),
+            )),
+
+            // For everything else, assume it's a permanent error and don't retry.
+            code => {
+                error!(checkpoint, %code, "Permanent error, giving up!");
+                Err(BE::permanent(Error::HttpError(checkpoint, code)))
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+pub(crate) mod tests {
+    use super::*;
+    use crate::ingestion::client::IngestionClient;
+    use crate::metrics::tests::test_metrics;
+    use axum::http::StatusCode;
+    use rand::{rngs::StdRng, SeedableRng};
+    use std::sync::Mutex;
+    use sui_storage::blob::{Blob, BlobEncoding};
+    use sui_types::full_checkpoint_content::CheckpointData;
+    use sui_types::{
+        crypto::KeypairTraits,
+        gas::GasCostSummary,
+        messages_checkpoint::{
+            CertifiedCheckpointSummary, CheckpointContents, CheckpointSummary,
+            SignedCheckpointSummary,
+        },
+        supported_protocol_versions::ProtocolConfig,
+        utils::make_committee_key,
+    };
+    use tokio_util::sync::CancellationToken;
+    use wiremock::{
+        matchers::{method, path_regex},
+        Mock, MockServer, Request, Respond, ResponseTemplate,
+    };
+
+    const RNG_SEED: [u8; 32] = [
+        21, 23, 199, 200, 234, 250, 252, 178, 94, 15, 202, 178, 62, 186, 88, 137, 233, 192, 130,
+        157, 179, 179, 65, 9, 31, 249, 221, 123, 225, 112, 199, 247,
+    ];
+
+    pub(crate) async fn respond_with(server: &MockServer, response: impl Respond + 'static) {
+        Mock::given(method("GET"))
+            .and(path_regex(r"/\d+.chk"))
+            .respond_with(response)
+            .mount(server)
+            .await;
+    }
+
+    pub(crate) fn status(code: StatusCode) -> ResponseTemplate {
+        ResponseTemplate::new(code.as_u16())
+    }
+
+    pub(crate) fn test_checkpoint_data(cp: u64) -> Vec<u8> {
```

Comment:
nit: This is quite a universal test helper, maybe this belongs in a `cfg(test)` module on the outer `ingestion` module?